### PR TITLE
Feature/trans invadj ec

### PR DESCRIPTION
**Pull request by MarekWlasak**
This is the associated PR that allows the adjoint of the inverse spectral transform to be used in atlas. 

It also includes a test that simulates a spectral horizontal background error covariance model and provides an adjoint test.

Code depends on an update to transi.

### CODE REVIEW

File: src/atlas/meshgenerator/detail/StructuredMeshGenerator.cc [hunk 1]
```diff
@@ -65,6 +65,10 @@ struct Region {
 };
 
 StructuredMeshGenerator::StructuredMeshGenerator( const eckit::Parametrisation& p ) {
+
+    Log::info() << "StructuredMeshGenerator::config parameterisation" <<
+       dynamic_cast<const eckit::Configuration &>(p) << std::endl;
+
```
Comment (wdeconinck) [inline src/atlas/meshgenerator/detail/StructuredMeshGenerator.cc:71]: Please remove this statement, it is too verbose for many applications.

File: src/atlas/meshgenerator/detail/StructuredMeshGenerator.cc [hunk 2]
```diff
@@ -142,6 +138,16 @@ StructuredMeshGenerator::StructuredMeshGenerator( const eckit::Parametrisation&
         }
         options.set( "partitioner", partitioner );
     }
+     else if ( grid::Partitioner::exists( "trans" ) ) {
+        partitioner = "trans";
+    } else {
+        partitioner = "equal_regions";
+    }
+    std::cout << "StructuredMeshGenerator::partitioner = " << partitioner << std::endl;
```
Comment (wdeconinck) [inline src/atlas/meshgenerator/detail/StructuredMeshGenerator.cc:146]: Please remove above print statement.

File: src/atlas/meshgenerator/detail/StructuredMeshGenerator.cc [hunk 3]
```diff
@@ -142,6 +138,16 @@ StructuredMeshGenerator::StructuredMeshGenerator( const eckit::Parametrisation&
...
+    options.set( "partitioner", partitioner );
```
Comment (wdeconinck) [inline src/atlas/meshgenerator/detail/StructuredMeshGenerator.cc:147]: Given that this statement is now here, the one at (new) line 139 can be removed.

File: src/atlas/trans/Trans.cc [hunk 4]
```diff
@@ -125,6 +125,28 @@ void Trans::invtrans_vordiv2wind( const Field& spvor, const Field& spdiv, Field&
...
+
+void Trans::invtrans_adj_grad( const FieldSet& gradfields, FieldSet& spfields, const eckit::Configuration& config ) const {
+    get()->invtrans_adj_grad( gradfields, spfields, options( config ) );
+}
+
+void Trans::invtrans_adj_vordiv2wind( const Field& gpwind, Field& spvor, Field& spdiv,
+                                     const eckit::Configuration& config ) const {
+    get()->invtrans_adj_vordiv2wind( gpwind, spvor, spdiv, options( config ) );
+}
+
```
Comment (wdeconinck) [inline src/atlas/trans/Trans.cc:149]: Should we not rather have the "_adj" as a function suffix rather than in the middle?

**Review by wdeconinck — COMMENTED** [review]
Thanks Marek, looking good with some minor cosmetic changes required.

I will just have to have a little think on how to have Atlas play nice with previous versions of transi (where the adjoint interface did not exist), but not your concern.

File: src/atlas/meshgenerator/detail/StructuredMeshGenerator.cc (hunk 1 above)
Comment (MarekWlasak) [reply src/atlas/meshgenerator/detail/StructuredMeshGenerator.cc]: Done!

File: src/atlas/meshgenerator/detail/StructuredMeshGenerator.cc (hunk 2 above)
Comment (MarekWlasak) [reply src/atlas/meshgenerator/detail/StructuredMeshGenerator.cc]: Done!

File: src/atlas/meshgenerator/detail/StructuredMeshGenerator.cc (hunk 3 above)
Comment (MarekWlasak) [reply src/atlas/meshgenerator/detail/StructuredMeshGenerator.cc:141]: Done!

File: src/atlas/trans/Trans.cc (hunk 4 above)
Comment (MarekWlasak) [reply src/atlas/trans/Trans.cc:149]: Done!

Comment (MarekWlasak) [reply src/atlas/trans/Trans.cc:149]: Done

File: src/atlas/meshgenerator/detail/StructuredMeshGenerator.cc (hunk 2 above)
Comment (MarekWlasak) [reply src/atlas/meshgenerator/detail/StructuredMeshGenerator.cc]: Done

File: src/atlas/meshgenerator/detail/StructuredMeshGenerator.cc (hunk 3 above)
Comment (MarekWlasak) [reply src/atlas/meshgenerator/detail/StructuredMeshGenerator.cc:141]: Done

**Comment by MarekWlasak** [comment]
Retry DeepCode

**Comment by MarekWlasak** [comment]
Hi Willem. I have applied all the changes that you have suggested. For some reason I am having a problem with the deepcode bot.

**Comment by MarekWlasak** [comment]
Retry DeepCode

**Comment by wdeconinck** [comment]
This is now merged in develop manually. Thanks @MarekWlasak !

**Pull request closed without merging**